### PR TITLE
fix(components): [splitter] preserve panel sizes across collapse and resize

### LINKED ISSUE #23015: [Component] [splitter] Incorrect panel sizes after container resize following collapse/expand
Bug Type: **`Component`**

## Environment

- Vue Version: `3.5.25`
- Element Plus Version: `2.12.0`
- Browser / OS: `Chrome 142.0.7444.176 /  MacOS Sequoia 15.7.1`
- Build Tool: `Vite`

### Related Component

- `el-splitter`
- `el-splitter-panel`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2XG4gICAgc3R5bGU9XCJoZWlnaHQ6IDI1MHB4OyBib3gtc2hhZG93OiB2YXIoLS1lbC1ib3JkZXItY29sb3ItbGlnaHQpIDBweCAwcHggMTBweFwiXG4gID5cbiAgICA8ZWwtc3BsaXR0ZXIgbGF6eT5cbiAgICAgIDxlbC1zcGxpdHRlci1wYW5lbCBjb2xsYXBzaWJsZSBtaW49XCI1MFwiPlxuICAgICAgICA8ZGl2IGNsYXNzPVwiZGVtby1wYW5lbFwiPjE8L2Rpdj5cbiAgICAgIDwvZWwtc3BsaXR0ZXItcGFuZWw+XG4gICAgICA8ZWwtc3BsaXR0ZXItcGFuZWwgY29sbGFwc2libGU+XG4gICAgICAgIDxkaXYgY2xhc3M9XCJkZW1vLXBhbmVsXCI+MjwvZGl2PlxuICAgICAgPC9lbC1zcGxpdHRlci1wYW5lbD5cbiAgICAgIDxlbC1zcGxpdHRlci1wYW5lbCBjb2xsYXBzaWJsZT5cbiAgICAgICAgPGRpdiBjbGFzcz1cImRlbW8tcGFuZWxcIj4zPC9kaXY+XG4gICAgICA8L2VsLXNwbGl0dGVyLXBhbmVsPlxuICAgIDwvZWwtc3BsaXR0ZXI+XG4gIDwvZGl2PlxuPC90ZW1wbGF0ZT5cblxuPHN0eWxlIHNjb3BlZD5cbi5kZW1vLXBhbmVsIHtcbiAgZGlzcGxheTogZmxleDtcbiAgYWxpZ24taXRlbXM6IGNlbnRlcjtcbiAganVzdGlmeS1jb250ZW50OiBjZW50ZXI7XG4gIGhlaWdodDogMTAwJTtcbn1cbjwvc3R5bGU+XG4iLCJlbGVtZW50LXBsdXMuanMiOiJpbXBvcnQgRWxlbWVudFBsdXMgZnJvbSAnZWxlbWVudC1wbHVzJ1xuaW1wb3J0IHsgZ2V0Q3VycmVudEluc3RhbmNlIH0gZnJvbSAndnVlJ1xuXG5sZXQgaW5zdGFsbGVkID0gZmFsc2VcbmF3YWl0IGxvYWRTdHlsZSgpXG5cbmV4cG9ydCBmdW5jdGlvbiBzZXR1cEVsZW1lbnRQbHVzKCkge1xuICBpZiAoaW5zdGFsbGVkKSByZXR1cm5cbiAgY29uc3QgaW5zdGFuY2UgPSBnZXRDdXJyZW50SW5zdGFuY2UoKVxuICBpbnN0YW5jZS5hcHBDb250ZXh0LmFwcC51c2UoRWxlbWVudFBsdXMpXG4gIGluc3RhbGxlZCA9IHRydWVcbn1cblxuZXhwb3J0IGZ1bmN0aW9uIGxvYWRTdHlsZSgpIHtcbiAgY29uc3Qgc3R5bGVzID0gWydodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL2VsZW1lbnQtcGx1c0BsYXRlc3QvZGlzdC9pbmRleC5jc3MnLCAnaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L3RoZW1lLWNoYWxrL2RhcmsvY3NzLXZhcnMuY3NzJ10ubWFwKChzdHlsZSkgPT4ge1xuICAgIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSwgcmVqZWN0KSA9PiB7XG4gICAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgICAgbGluay5ocmVmID0gc3R5bGVcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignbG9hZCcsIHJlc29sdmUpXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2Vycm9yJywgcmVqZWN0KVxuICAgICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgICB9KVxuICB9KVxuICByZXR1cm4gUHJvbWlzZS5hbGxTZXR0bGVkKHN0eWxlcylcbn1cbiIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AdnVlL3NoYXJlZEBsYXRlc3QvZGlzdC9zaGFyZWQuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC9kaXN0L2luZGV4LmZ1bGwubWluLm1qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzL1wiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC9cIixcbiAgICBcIkBlbGVtZW50LXBsdXMvaWNvbnMtdnVlXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AZWxlbWVudC1wbHVzL2ljb25zLXZ1ZUAyL2Rpc3QvaW5kZXgubWluLmpzXCJcbiAgfSxcbiAgXCJzY29wZXNcIjoge31cbn0iLCJfbyI6e319)

### Steps to reproduce

Example: Initial state Container 1000px, Panel A 500px, Panel B 500px
1. Collapse Panel A → cacheCollapsedSize[0] = 500px
2. Container resizes to 800px → Panel A 0px, Panel B 800px
3. Expand Panel A → Restores to 500px (WRONG! Should be 400px)

https://github.com/user-attachments/assets/efad8b24-c6c2-49a4-b4bf-64941c6c4fe7

### What is Expected?

Panel A 400px (50%), Panel B 400px (50%)

### What is actually happening?

Panel A 500px (62.5%), Panel B 300px (37.5%)

### Additional comments

I have already implemented and tested this fix locally. I will submit a PR soon.

### PR DESCRIPTION
**Pull request by hnviradiya**
- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

Fixes #23015
Refs #22944

> **Prior art — please read first.**
>
> @tuzixiangs' #23016 diagnosed this root cause first and fixes it in 18 lines; it is approved and predates this PR by some months. I opened this before finding it. Credit for the diagnosis is theirs, and I'm glad to close this in favour of #23016, or to have this work folded into it — whichever the maintainers prefer.
>
> What this adds on top is the part #23016 stalled on: @tuzixiangs [reported](https://github.com/element-plus/element-plus/pull/23016#issuecomment-3677551026) being unable to make `px`-sized panels hold their width, listing three obstacles. Those are solved here, along with the "Option 2" unit semantics @btea signed off on, and the documentation that was asked for.

## Problem

Two distinct bugs, both visible after collapsing and expanding a panel.

**1. Proportions are lost when the container resizes (#23015).** The collapse/expand cache stored each panel's size as a raw pixel value, so a container resize while a panel sat collapsed restored a stale width. Container 1000px, two 50/50 panels: collapse A, resize to 800px, expand A → 500/300 instead of 400/400.

**2. Explicitly sized panels drift (#22944).** Collapsing and expanding writes live pixel numbers into every panel's internal size, which erases the distinction between a panel that asked for a fixed width and one that is simply sharing what's left. From that point on, resizing the container rescales everything: the reporter's 1st and 3rd panels stop holding their widths and all three panels move together.

## Fix

**The cache** now records *how* a size was expressed, not just its pixel value, and resolves it against the container at the moment of expanding — a ratio for proportional panels, a literal width for `px` ones. It is keyed per panel rather than per bar, so a middle panel collapsed from one bar comes back correctly when expanded from the other, and expanding applies the panel limits in the same order `onMoving` does, so a restored size can't land outside the limits a drag would enforce.

**The size unit is the signal of intent**, as agreed in #23016:

| `size` | Behaviour |
| --- | --- |
| `"200px"` | Stays 200px. The rest of the container is shared out among the other panels. |
| `:size="200"` | 200px to begin with, then proportional. |
| `"25%"` | Always a quarter of the container. |
| omitted | Shares what the sized panels leave over. |

Making that real for `px` panels needed an answer to each obstacle raised on #23016:

- **Fixed widths exceeding a shrunken container** — the pinned panels scale down together rather than pushing the flexible ones below zero. 250px of pins in a 200px container renders 80/0/120.
- **The cache conflicting with allocation** — the pin comes from the *declared* width, never the live `size`, so a panel that absorbed a collapsed neighbour is not re-pinned at the absorbed width. Only a panel the user actually collapsed pins at 0; one merely squeezed to 0 by a neighbour's expand keeps its pin and recovers.
- **Drag boundaries** — dragging a pinned panel is a deliberate resize, so it redefines the pinned width. Collapse and expand deliberately do not.

`min`/`max` keep their existing unit rules — a percentage tracks the container, a number or `px` string is a literal pixel limit — since that is how `getLimitSize` has always read them.

## Testing

`pnpm vitest run packages/components/splitter` — 34 passing. Twelve regression tests added, each verified to fail without its corresponding fix, covering: the #23015 root cause; `px` panels holding their width through a collapse cycle and a container resize; bare numbers and numeric strings staying proportional; pinned panels scaling down on overflow; a squeezed panel recovering; per-panel cache keying for middle panels; restores honouring the panel's own and its neighbour's limits; `v-model` echoes not being mistaken for authored sizes, and a genuinely later change not being mistaken for an echo; and `null`/`''`/`NaN` values not being coerced into limits or sizes.

Behaviour was also swept against `origin/dev` rather than only unit-tested — size modes × collapse direction × panel position × limit forms (number, bare string, `px`, `%`, `null`, `''`, non-numeric, on the panel and its neighbour) × container grow/shrink/overflow/zero × `v-model` × dynamic add and remove × vertical layout × drag-then-collapse. Every difference from `dev` is an intended one.

Locally reproducing CI: `pnpm lint` clean, `pnpm typecheck` clean across all five projects, `pnpm test` 2563 passing.

## Documentation

`docs/en-US/component/splitter.md` gains a "Size units" section describing the table above, per @btea's request on #23016 that the difference be documented.

## Known limitation, unchanged from `dev`

At a zero-width container the collapse icons disappear, so a collapsed panel can't be expanded until the container has a size again.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved splitter panel collapse and expansion across container resizing.
  * Preserved fixed-pixel and proportional panel sizes more reliably.
  * Improved handling of minimum and maximum constraints, overflow, and flexible space.
  * Corrected behavior for numeric-string, empty, and unset size limits.

* **Documentation**
  * Added guidance on splitter size units and collapse/resize behavior.

### CODE REVIEW

File: packages/components/splitter/src/split-panel.vue [hunk 1]
```diff
@@ -112,11 +112,24 @@ function sizeToPx(str: string | number | undefined) {
...
+
 // Two-way binding for size
 let isSizeUpdating = false
 watch(
   () => props.size,
   () => {
+    if (props.size === lastEmittedSize) {
+      lastEmittedSize = undefined
+    } else {
+      declaredSize.value = props.size
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/split-panel.vue]: Confirmed and fixed in `ce1b146`.

The echo marker is now scoped to a single tick: `rememberEmittedSize()` records the value and clears it on `nextTick`, mirroring how `isSizeUpdating` already works. A parent mirroring `update:size` writes back within the same flush and is still recognised as the echo, while a debounced, delayed, or entirely ignored emit leaves nothing behind — so a later `:size` change to that same number is treated as newly authored.

Regression test `should treat a later size prop change as authored, not as an echo` covers exactly the sequence you described: a one-way `:size="'50%'"` binding whose `update:size` emits are discarded, then an explicit `size = 500`, then collapse → resize to 800px → expand. It restores 500px as the authored fixed size; on the previous commit it restored 400px as a 50% ratio.

File: packages/components/splitter/src/hooks/useResize.ts [hunk 2]
```diff
@@ -128,13 +128,53 @@ export function useResize(
...
-        ...pxSizes.value.map((size, i) =>
-          size <= 0 ? getLimitSize(limitSizes.value[i]?.[0], 0) : size
-        )
+        ...pxSizes.value.map((size, i) => {
+          const minLimit = limitSizes.value[i]?.[0]
+          const validSize = size <= 0 ? getLimitSize(minLimit, 0) : size
+          return toCachedSize(
+            validSize,
+            isFixedForCache(panels.value[i], minLimit)
+          )
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/hooks/useResize.ts:187]: Fixed in `ce1b146`, though the diagnosis is worth stating precisely: this was a type-soundness problem rather than a behavioural one.

`isPct()` and `isPx()` are `itemSize is string` predicates, so their false branches narrow `string | number | undefined` down to `number | undefined` — which is why `getLimitSize` was *inferred* as returning `number` while a bare numeric string (`min="100"`, as a Vue template passes it) still flowed through as a runtime string. Nothing downstream actually misbehaved: `Math.max`, `Math.min` and lodash's `clamp` all coerce, and the ratio branch divides, so every consumer produced the right number anyway.

Rather than normalising at the call site, I made `getLimitSize` return a real `number` — explicit `: number` annotation, `Number()` conversion for the bare-string case, and the `defaultLimit` fallback for anything non-numeric. That fixes it at the source, so the `Number()` wrappers I had at the limit call sites are gone too and `CachedSize.value` can no longer receive a string.

Since `getLimitSize` is shared with `onMoving`, I diffed drag and collapse behaviour with and without this change across every limit form — `min`/`max` as number, bare string, `px` string and `%` string, on the panel itself and on its neighbour, plus both-limits-set — 24 scenarios, byte-identical output either way. `should honour a bare numeric string min when restoring` is added as a guard; being honest, it passes on the previous commit too, so it documents the contract rather than catching a live bug.

File: packages/components/splitter/src/hooks/useResize.ts [hunk 3]
```diff
@@ -17,13 +17,20 @@ export function useResize(
...
     defaultLimit: number
-  ) {
+  ): number {
     if (isPct(str)) {
       return ptg2px(getPct(str))
     } else if (isPx(str)) {
       return getPx(str)
     }
-    return str ?? defaultLimit
+    if (str === undefined) {
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/hooks/useResize.ts]: Correct, and this one was a regression I introduced — thank you. Fixed in `0626a30`.

`getLimitSize` now returns `defaultLimit` for `null` as well as `undefined` (`str == null`), so a limit bound to `null` is unset again rather than becoming a 0px maximum.

Re-sweeping the limit forms against `origin/dev` afterwards turned up a second instance of the same class, which your report led me to: `max=""` also reached `Number('') === 0`. On `dev` that quirk is invisible on expand (the expand path applied no limits at all) but very visible on drag — a `max=""` panel is pinned shut at `1000/0`. Since this PR makes expand honour the same limits as dragging, the quirk started showing up there too. `''` is now treated as unset alongside `null`.

That does mean one deliberate divergence from `dev`: dragging a `max=""` panel now behaves as unlimited instead of pinning it to 0. It's an incidental fix for what looks like an existing bug, and I'd rather flag it explicitly than let it ride silently — happy to drop it back to the previous behaviour if you'd prefer this PR stay strictly scoped.

Verification: drag and collapse/resize/expand behaviour diffed against `origin/dev` across 18 limit forms — `min`/`max` as number, bare numeric string, `px`, `%`, `null`, `''`, non-numeric, `undefined`, on the panel and on its neighbour, and both set at once. Drag output is identical to `dev` on every form except the `max=""` case above; the expand rows all land on the corrected proportional restore with no outliers. Tests `should treat a null limit as unset` and `should treat an empty string limit as unset` both fail on the previous commit (`800/0`) and pass now.

File: packages/components/splitter/src/split-panel.vue [hunk 4]
```diff
@@ -112,11 +112,31 @@ function sizeToPx(str: string | number | undefined) {
...
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/split-panel.vue]: Confirmed and fixed in `0626a30`.

You're right that comparing against `lastEmittedSize` alone is ambiguous: at rest it is `undefined`, so a parent clearing `size` back to `undefined` matched the echo branch and `declaredSize` kept its old fixed value. There is now an explicit `hasPendingEcho` flag that `rememberEmittedSize()` sets and the same-tick `nextTick` clears, and the echo branch requires it — so the check only fires when an emit is genuinely outstanding.

Regression test `should treat a cleared size prop as a real change, not an echo`: a panel declared `size={150}`, cleared to `undefined`, then dragged to 400px, collapsed, container shrunk to 800px, expanded. It now restores 320px as the auto-sized panel it became; on the previous commit it restored the stale fixed 400px.

File: packages/components/splitter/src/hooks/useSize.ts [hunk 5]
```diff
@@ -24,6 +24,32 @@ export function isPx(
...
+}
+
+// Whether a *declared* size prop always resolves to exactly 0 (e.g. "0%",
+// "0px", or 0), regardless of container size - unlike undefined/'' (auto-fill),
+// which depends on siblings and isn't deterministically zero.
+export function isZeroSize(itemSize: string | number | undefined): boolean {
+  if (itemSize === undefined || itemSize === '') return false
+  if (isPct(itemSize)) return getPct(itemSize) === 0
+  if (isPx(itemSize)) return getPx(itemSize) === 0
+  const num = Number(itemSize)
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/hooks/useSize.ts]: Confirmed and fixed in `90ea0b6` — and you caught me mid-audit on exactly this. After the `getLimitSize` null regression I started checking every value form these classifiers accept against how the `useSize` loop treats the same value, and `null` was the first divergence the table showed:

| `size` | `isFixedSize` | `isZeroSize` | `useSize` loop |
| --- | --- | --- | --- |
| `undefined` / `''` | false | false | auto-fill |
| **`null`** | **true** | **true** | **auto-fill** |
| **`NaN`** | **true** | false | **auto-fill** |
| `0` / `'0'` / `'0px'` | true | true | pixel |
| `'0%'` | false | true | percent |
| `150` / `'150'` / `'150px'` | true | false | pixel |
| `'50%'` | false | false | percent |
| `'abc'` | false | false | auto-fill |

So there were two rows out of step, not one — `NaN` was wrong for the same underlying reason, via the `typeof itemSize === 'number'` clause.

Both classifiers now bail out on `itemSize == null` alongside `undefined`/`''`, and the `typeof` clause is gone in favour of `!Number.isNaN(Number(itemSize))`, which makes `NaN` fall through to auto-fill exactly as the loop does. Every row in that table now matches the loop.

Regression test `should treat a null size as auto-fill, not as a fixed pixel size` covers the visible symptom: `<ElSplitterPanel :size="null" :min="100">` in a 1000px container, collapse, shrink to 800px, expand. It restores 400px as an auto-filled panel now; on the previous commit `isZeroSize(null)` was true so the panel was classified through its `min` units and pinned at a stale 500px.

File: docs/en-US/component/splitter.md [hunk 6]
```diff
@@ -57,6 +57,25 @@ splitter/size
...
+| `size`        | Behaviour                                                                                 |
+| ------------- | ----------------------------------------------------------------------------------------- |
+| `"200px"`     | Stays 200px wide. The remaining space is shared out among the other panels.               |
+| `:size="200"` | 200px to begin with, then keeps its share of the container proportionally.                |
+| `"25%"`       | Always a quarter of the container.                                                        |
+| omitted       | Shares whatever space the sized panels leave over, equally with the other omitted panels. |
+
+A panel sized in `px` keeps that width through collapse and expand as well.
+When the `px` panels together are wider than the container, they shrink in
+proportion rather than pushing the other panels out.
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply docs/en-US/component/splitter.md]: Good catch — the table was overstating things. Reworded in `8e26b67`.

The column is now "Panel aims for" rather than a flat statement of what happens, and the section spells out that these are targets rather than guarantees: `min` and `max` take precedence, a panel can be collapsed to zero whatever its size says, and `px` panels shrink in proportion when their combined width exceeds the container.

File: docs/en-US/component/splitter.md [hunk 7]
```diff
@@ -57,6 +57,25 @@ splitter/size
 
 :::
 
+### Size units
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply docs/en-US/component/splitter.md:60]: Skipping this one deliberately, and flagging it for a maintainer rather than guessing.

`size`, `min` and `max` are not new APIs, so the section isn't documenting a new prop — what is new is the behaviour that a `px` size stays pinned. A `^(x.y.z)` tag would therefore be marking a behaviour change, and I have no way to know which release that lands in: doc tags in `dev` currently run to `^(2.14.4)`, and the workspace version is the `0.0.0-dev.1` placeholder, so anything I wrote would be a guess that someone has to correct later.

Happy to add whichever version you intend this for — just say the word and I'll push it.

File: packages/components/splitter/src/hooks/useSize.ts [hunk 8]
```diff
@@ -37,11 +60,21 @@ export function useSize(
...
     // Convert the passed props size to a percentage
     for (let i = 0; i < panelCounts.value; i += 1) {
       const itemSize = panels.value[i]?.size
 
+      // A pinned panel the user collapsed holds at 0 - re-pinning it to its
+      // declared width would make it impossible to collapse. One merely
+      // squeezed to 0 by a neighbour keeps its pin and recovers.
+      fixedPx[i] = panels.value[i]?.isCollapsed
+        ? 0
+        : panels.value[i]?.fixedPxSize
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/hooks/useSize.ts]: Confirmed and fixed in `6ae6ccc`.

Pinning at 0 now requires both conditions - the panel is flagged as collapsed *and* its live size is still 0. The flag alone ignored a parent writing a real size back, exactly as you describe; a zero size alone would strand a panel that a neighbour's expand merely squeezed to 0, which is why the flag exists. Requiring both covers each case. `split-panel.vue` also clears the flag outright when a non-zero `size` arrives from the parent, so the state stays honest rather than just being masked at render time.

Regression test `should honour a size prop written while a px panel is collapsed`: a `size="200px"` panel collapsed via the control, then driven to `"300px"` by the parent. It renders 300px now; before the fix it stayed pinned at 0.

File: packages/components/splitter/src/split-panel.vue [hunk 9]
```diff
@@ -165,6 +201,17 @@ const _panel = reactive({
...
+  // Tied to the declared size (not the internal `size` below, which useResize
+  // overwrites with a live raw px number while dragging/collapsing) so a
+  // panel's declared size mode survives those internal mutations.
+  isFixedSize: computed(() => isFixedSize(declaredSize.value)),
+  // Seeded from the declared "Npx" width and kept up to date by drags (see
+  // useResize) - collapsing a neighbour must not redefine what this panel is
+  // pinned to, but the user dragging it deliberately must.
+  fixedPxSize: isFixedSize(props.size)
+    ? getPx(props.size as string)
+    : undefined,
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/split-panel.vue]: Confirmed and fixed in `6ae6ccc`.

`"0px"` is where the panel starts rather than a width to hold it at, so it is no longer given a pinned width at all - the same treatment the other zero-size declarations already had. `min` is what it expands to, and the collapse cache was already restoring that correctly; the stale `fixedPxSize: 0` was undoing it on the next recomputation.

Regression test `should expand a "0px" panel to its min rather than pinning it shut`: `<ElSplitterPanel size="0px" :min="100">` expands to 100px now, where before it snapped back to 0.

Worth noting the fix initially looked like it had landed when it hadn't - the edit was silently reformatted away, and the test still failed. Reading back `fixedPxSize` from the live panel state is what showed it was still `0`, which is also what confirms the fix now takes effect.

File: packages/components/splitter/src/split-panel.vue [hunk 10]
```diff
@@ -112,11 +117,46 @@ function sizeToPx(str: string | number | undefined) {
...
 watch(
   () => props.size,
   () => {
+    if (hasPendingEcho && props.size === lastEmittedSize) {
+      hasPendingEcho = false
+      lastEmittedSize = undefined
+    } else {
+      declaredSize.value = props.size
+      if (panel.value) {
+        panel.value.fixedPxSize = pinnedPxSize(props.size)
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/split-panel.vue:163]: Confirmed and fixed in `bf8c5bc`.

The pin is now taken from the width *after* `min`/`max` rather than the authored one, so it can't smuggle a value past the limits the size watcher applies. The clamp is shared between the two places that record a pin — the initial registration and the prop-change path — which also closes the same hole on first render, where the watcher hasn't run yet.

Regression test `should clamp a pinned px size to the panel limits`: a panel with `max="300px"` driven from `"200px"` to `"500px"` renders 300px and holds 300px across a container resize. On the previous commit it rendered 500px.

File: packages/components/splitter/src/hooks/useResize.ts [hunk 11]
```diff
@@ -109,6 +121,11 @@ export function useResize(
     updatePanelSizes = () => {
       panels.value.forEach((panel, index) => {
         panel.size = numSizes[index]
+        // Dragging a pinned panel is a deliberate resize, so it becomes the
+        // new pinned width. Collapse/expand deliberately does not do this.
+        if (panel.isFixedSize && numSizes[index]) {
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/hooks/useResize.ts]: Confirmed and fixed in `bf8c5bc`.

The guard is now an explicit `!== undefined` rather than a truthiness check, so a drag to 0 records 0 as the new pinned width instead of leaving the previous one in place to undo it on the next recomputation.

Regression test `should let a px panel be dragged to zero and stay there`: a `size="200px"` panel dragged to 0 stays at 0 through a container resize; on the previous commit it snapped back to 200px.

File: packages/components/splitter/src/hooks/useResize.ts [hunk 12]
```diff
@@ -109,6 +121,13 @@ export function useResize(
     updatePanelSizes = () => {
       panels.value.forEach((panel, index) => {
         panel.size = numSizes[index]
+        // Dragging a pinned panel is a deliberate resize, so it becomes the
+        // new pinned width - including a drag right down to 0, which a
+        // truthiness check would skip and then immediately undo. Collapse and
+        // expand deliberately do not update the pin.
+        if (panel.isFixedSize && numSizes[index] !== undefined) {
+          panel.fixedPxSize = numSizes[index]
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/hooks/useResize.ts]: Confirmed and fixed in `2a3a894` — and this one was a regression I introduced in the previous commit, so thank you for catching it quickly.

Loosening the guard to `!== undefined` (so a deliberate drag to 0 would stick) meant *every* panel at 0 recorded a pin of 0 on *any* drag, including one collapsed elsewhere. The update is now restricted to the two panels either side of the bar being dragged, which is the only place a drag can change a width anyway.

Regression test `should not pin a collapsed px panel shut when another bar is dragged`: a `size="200px"` panel is collapsed, the *second* bar is then dragged, and the panel is expanded again — it comes back to 200px and holds it across a container resize. On the previous commit it was stuck at 0px.

File: packages/components/splitter/src/hooks/useSize.ts [hunk 13]
```diff
@@ -70,9 +106,53 @@ export function useSize(
...
+    // would leave part of the container empty, so they share it out as before
+    // and the pinned width comes back from the collapse cache on expand.
+    if (!restPtg) return ptgList
+
+    const restShare = (containerSize.value - fixedTotal) / containerSize.value
+
+    return ptgList.map((ptg, i) =>
+      fixedPx[i] === undefined
+        ? (ptg / restPtg) * restShare
+        : fixedPx[i]! / containerSize.value
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/hooks/useSize.ts:212]: Confirmed and fixed in `2a3a894`.

Pins are now re-clamped against the limits resolved at the *current* container size on every pass, rather than the stored pixel value being reused as-is, so a percentage limit is honoured as the container changes. A pin of 0 is left alone, since that means a collapse or a deliberate drag shut rather than a width to clamp.

This needed the limit resolver in both hooks, so `getLimitSize` has moved into `useSize` as an exported `resolveLimit(limit, containerSize, fallback)`, with `useResize` keeping a thin wrapper that binds the current container size. Same logic, one copy — the `%`/`px`/number/`null`/`''`/non-numeric handling is unchanged.

Regression test `should re-clamp a pinned px size against percentage limits`: `size="500px" max="50%"` renders 500px in a 1000px container and 400px once it shrinks to 800px. On the previous commit it stayed at 500px.

File: packages/components/splitter/src/hooks/useSize.ts [hunk 14]
```diff
@@ -70,9 +121,69 @@ export function useSize(
...
+
+    // Nothing left to share: fixed widths alone overflow the container, so
+    // they scale down together rather than pushing the others below zero.
+    if (fixedTotal >= containerSize.value) {
+      const scale = fixedTotal ? containerSize.value / fixedTotal : 0
+      return ptgList.map((ptg, i) =>
+        fixedPx[i] === undefined
+          ? 0
+          : (fixedPx[i]! * scale) / containerSize.value
+      )
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/hooks/useSize.ts:169]: Confirmed and fixed in `3f81507`.

The overflow path was scaling every pin by the same factor, which spent a `min` that was still affordable. Each pinned panel now gives up only what it has *above its own min*, in proportion to that room, so a roomier neighbour absorbs the shortfall first. If even the minimums don't fit, they shrink together as before — there's nothing left to honour at that point.

Your example resolves as you'd expect: `size="200px" min="200px"` and `size="900px"` in a 1000px container render 200px and 800px, instead of 181.8px and 818.2px.

Regression test `should keep a satisfiable min when pinned panels overflow` covers exactly that layout. The existing overflow test (three panels with no `min`, 250px of pins in a 200px container) still lands on the same 80/0/120 as before, since with no floors the two behave identically.

File: packages/components/splitter/src/hooks/useResize.ts [hunk 15]
```diff
@@ -128,13 +127,62 @@ export function useResize(
...
+    cached ? (cached.isRatio ? ptg2pxSize(cached.value) : cached.value) : 0
+
+  // Limits follow different unit rules to sizes: `getLimitSize` reads a bare
+  // number or an "Npx" string as a literal pixel floor, and only a '%' string
+  // scales with the container.
+  const isFixedLimit = (limit: string | number | undefined) =>
+    limit != null &&
+    limit !== '' &&
+    !isPct(limit) &&
+    !Number.isNaN(Number(limit))
```
[the remark this answers is not in the document]
Comment (hnviradiya) [reply packages/components/splitter/src/hooks/useResize.ts]: Confirmed and fixed in `8e47361` — apologies for the slow reply, this one sat while I was away from the PR.

You're right that the two functions disagreed: `resolveLimit` reads `"100px"` as a literal width, while this classifier fell through to `Number('100px')`, got `NaN`, and called it proportional. It now checks `isPx(limit)` before the numeric test, so a pixel-string limit is treated as fixed exactly like a bare number.

Regression test `should restore a collapsed-to-zero panel's px-string min after container resize`: `size="0%" min="100px"` seeded at a 400px container, container grown to 800px while the panel is still at 0, then expanded. It restores the literal 100px now; on the previous commit it restored 200px, exactly as you described.

Worth recording that the fix initially broke five unrelated zero-size tests, all restoring 0. `isPx` had been dropped from this file's imports during an earlier tidy-up, so the new predicate was calling an undefined function and `onCollapse` threw. The suite caught it immediately, and it is a good argument for the wider net: the failure looked nothing like the change.
